src/Encoding.chpl

```diff
@@ -0,0 +1,255 @@
+/* Documentation for Encoding */
+module Encoding {
+
+  use SysCTypes;
+
+  // wireTypes
+  const varint = 0;
+  const lengthDelimited = 2;
+  const fixed64Type = 1;
+  const fixed32Type = 5;
+
+  proc unsignedVarintDump(val:uint, ch) throws {
+    if val == 0 {
+      ch.write(b"\x00");
+      return;
+    }
+  
+    var newVal = val;
+    var shiftVal: uint;
+    while newVal != 0 {
+      shiftVal = newVal >> 7;
+      var k = if shiftVal != 0 then 0x80 else 0x00;
+      var newByte = (newVal & 0x7F | k):uint(8);
+      ch.writeBytes(newByte, 1);
```

Comment:
This does not seem write to me. Anyway in general I don't think this code should be calling readBytes / writeBytes; rather the channel should be configured for binary I/O (and specify the appropriate endianness) and then this code should just do read and write.

Reply:
I guess you are pointing towards something like this ->
```
var f = open("out", iomode.cw, style = new iostyle(binary=1));
```
If our current implementation is seen, the channel passed to the serialize/unserialize procs is controlled by the user. Is there any other way we can achieve the binary configuration?

Reply:
Yes, there are two ways.

First, one can call 'set_style' - see https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L2443 and https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L2459 . You will want to lock the channel before doing these (and unlock it after) - see https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L2193 . I think it would make sense to lock the channel for the entire serialization / deserialization operation.

Second, it's possible to create a new channel based upon another existing channel. For example see https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L3162-L3169 . This is relevant because the Chapel channel types incudes a `param` (known at compile time) indicating binary I/O and endianness (see code starting at https://github.com/chapel-lang/chapel/blob/c54a5da1fec9f5d889a805c54485d728db6dc467/modules/standard/IO.chpl#L461 ) . So, it is possible to wrap the existing channel in a binary-big-endian channel, for example. This can be faster and allows us to skip even code handling locking. However there isn't a good user-facing way to do it yet (although I think such a thing would be reasonable to add to the I/O system; if we choose this path, the protobuf code will use some internal details of channel - or we have to add the user-facing thing to IO.chpl soon).

Reply:
Oh, another interesting thing about the second option, is that it would allow all of your functions working with channels to have a concrete type.

Reply:
A third option is to use `writef`/`readf`. In practice this will just modify the channel's style to indicate binary I/O for each operation. (You could do that also - modify the style per-operation).